tidyr - pivot_ instead of gather & spread (#684)
datacarpentry/R-ecology-lesson: 03-dplyr.Rmd

>   as rows, and the
>   number of genera per plot as the values. You will need to summarize before
>   reshaping, and use the function `n_distinct()` to get the number of unique
>   genera within a particular chunk of data. It's a powerful function! See
>   `?n_distinct` for more.
> 
>```{r, answer=TRUE, purl=FALSE}
>surveys_wide_genera <- surveys %>%
>   group_by(plot_id, year) %>%
>   summarize(n_genera = n_distinct(genus)) %>%
>   pivot_wider(names_from = year, values_from = n_genera)
> 
>head(surveys_wide_genera)
>```
>
> 2. Now take that data frame and `pivot_longer()` it again, so each row is a unique
>    `plot_id` by `year` combination.
>
> ```{r, answer=TRUE, purl=FALSE}
> surveys_wide_genera %>%
>   pivot_longer(names_to = "year", values_to = "n_genera", cols = -plot_id)
> ```
>
> 3. The `surveys` data set has
>    two measurement columns: `hindfoot_length` and `weight`.  This makes it
>    difficult to do things like look at the relationship between mean values of
>    each measurement per year in different plot types. Let's walk through a
>    common solution for this type of problem. First, use `pivot_longer()` to create a

> >     dataset where we have a key column called `measurement` and a

Reviewer: Maybe it would be clearer to write something like

> There is a column called `measurement` that contains the type of measurement (hindfoot length or `weight) ...

> >    `value` column that takes on the value of either `hindfoot_length` or

Reviewer: I thin the version in the PR is correct, i.e. , the name of the new column should be in backticks (and singular form), `value`. This is analogous to the previous sentence where the column name `measurement` is in backticks.

>    `weight`. *Hint*: You'll need to specify which columns are being gathered.
>
> ```{r, answer=TRUE, purl=FALSE}
>surveys_long <- surveys %>%
> pivot_longer(names_to = "measurement", values_to = "value", cols = c(hindfoot_length, weight))
>
> ```
>
> 4. With this new data set, calculate the average of each
>    `measurement` in each `year` for each different `plot_type`. Then
>    `pivot_wider()` them into a data set with a column for `hindfoot_length` and
>    `weight`. *Hint*: You only need to specify the key and value
>    columns for `pivot_wider()`.
>
> ```{r, answer=TRUE, purl=FALSE}
> surveys_long %>%
>   group_by(year, measurement, plot_type) %>%
>   summarize(mean_value = mean(value, na.rm=TRUE)) %>%
>   pivot_wider(names_from = measurement, values_from = mean_value)
> ```

```{r, eval=FALSE, purl=TRUE, echo=FALSE}

> ## Reshaping challenges

Reviewer 2: I will admit this part is somewhat confusing, as it is a repetition of rows 624 (referenced to the original) and up, but without code; however, it's not part of this PR so I'll refrain from further commenting. Happy to know what the idea is here though :)

Reviewer: - ln 603-655 are the text and solutions as they appear on the course website.
- ln 659-671 are in a special code chunk that will be extracted via knitr hook into a separate document (the "code handout"). The content of the chunk won't show up on the course website, only in the code handout (see the ``{r, eval=FALSE, purl=TRUE, echo=FALSE}`` ? )

I also found this vey confusing when I first went through the source. It is mentioned in the contributor notes, I believe. But most people don't read these.

It's also a hassle because it means keeping two long pieces of text in synchrony, and I'm not sure many people actually use the handout to teach this course. I'd say get rid of the code handout, but that's a different discussion and PR :)

## 1. Reshape the `surveys` data frame with `year` as columns, `plot_id` as rows, and the number of genera per plot as the values. You will need to summarize before reshaping, and use the function `n_distinct()` to get the number of unique genera within a particular chunk of data. It's a powerful function! See `?n_distinct` for more.

## 2. Now take that data frame and `pivot_longer()` it again, so each row is a unique `plot_id` by `year` combination.

> ## 3. The `surveys` data set has two measurement columns: `hindfoot_length` and `weight`. This makes it difficult to do things like look at the relationship between mean values of each measurement per year in different plot types. Let's walk through a common solution for this type of problem. First, use `pivot_longer()` to create a dataset where we have a key column called `measurement` and a `value` column that takes on the value of either `hindfoot_length` or `weight`. *Hint*: You'll need to specify which columns are being gathered.

Reviewer 2: ```suggestion
## 3. The `surveys` data set has two measurement columns: `hindfoot_length` and `weight`. This makes it difficult to do things like look at the relationship between mean values of each measurement per year in different plot types. Let's walk through a common solution for this type of problem. First, use `pivot_longer()` to create a dataset where we have a names column called `measurement` and a values column that takes on the value of either `hindfoot_length` or `weight`. *Hint*: You'll need to specify which columns will be part of the reshape.
```

Reviewer: my previous comment applies here, as the lesson and code handout  should be kept in synchrony.

> ## 4. With this new data set, calculate the average of each `measurement` in each `year` for each different `plot_type`. Then `pivot_wider()` them into a data set with a column for `hindfoot_length` and `weight`. *Hint*: You only need to specify the key and value columns for `pivot_wider()`.

Reviewer 2: ```suggestion
## 4. With this new data set, calculate the average of each `measurement` in each `year` for each different `plot_type`. Then `pivot_wider()` them into a data set with a column for `hindfoot_length` and `weight`. *Hint*: You only need to specify the names and values columns for `pivot_wider()`.
```

folder.

```{r, purl=FALSE, eval=FALSE}

> write_csv(surveys_complete, path = "data/surveys_complete.csv")

Reviewer 2: The `path` argument is deprecated and `file` is the replacement.
```suggestion
write_csv(surveys_complete, file = "data/surveys_complete.csv")
```

```

```{r, purl=FALSE, eval=TRUE, echo=FALSE}
if (!dir.exists("data")) dir.create("data")

> write_csv(surveys_complete, path = "data/surveys_complete.csv")

Reviewer 2: ```suggestion
write_csv(surveys_complete, file = "data/surveys_complete.csv")
```
